### PR TITLE
feat: Eliminate redundant configuration reads

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message

```
feat: Eliminate redundant configuration reads

When instance id hasn't changed and datasource hasn't changed, don't
forcibly reload configuration.
```

## Additional Context
On a test system, this avoids reloading the configuration on the second `instancify()`

See the new messages in logs: 
[no-reload.log](https://github.com/user-attachments/files/16351318/no-reload.txt)

### CODE REVIEW

**Comment by TheRealFalcon** [comment]
What happens on a new instance id? The old instance directory and link won't go away, so won't we miss removing and re-linking the new instance directory?

**Comment by holmanb** [comment]
> What happens on a new instance id? The old instance directory and link won't go away, so won't we miss removing and re-linking the new instance directory?

The remove and link code will still run if the directory that the symlink already resolves to is different from the one that we would otherwise create.

**Review by TheRealFalcon — COMMENTED** [review]
The changes look good, but given it's age and how core it is to how cloud-init functions, I really don't think I'm comfortable changing anything here unless we have tests around ways instance-ids may change. Specifically:
* booting with the first instance-id
* booting with a new instance-id from previous boot
* booting with the same instance when there's no way to verify the instance-id on disk
* booting with a new instance id when there's no way to verify the instance-id on disk

**Comment by holmanb** [comment]
> The changes look good, but given it's age and how core it is to how cloud-init functions, I really don't think I'm comfortable changing anything here unless we have tests around ways instance-ids may change. Specifically:
> 
>     * booting with the first instance-id

see the latest test addition

>     * booting with a new instance-id from previous boot

see the latest integration test addition

>     * booting with the same instance when there's no way to verify the instance-id on disk
>     * booting with a new instance id when there's no way to verify the instance-id on disk

there is a test already in tests/integration_tests/datasources/test_caching.py that covers this case

**Comment by holmanb** [comment]
Rebased to pick up the labeler fix.

File: tests/integration_tests/integration_settings.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
 # <image_id>[::<os>::<release>::<version>].  If given, os and release should
 # describe the image specified by image_id.  (Ubuntu releases are converted
 # to this format internally; in this case, to "None::ubuntu::focal::20.04".)
-OS_IMAGE = "focal"
+OS_IMAGE = "jammy"
```
Comment (TheRealFalcon) [inline tests/integration_tests/integration_settings.py:40]: Were these changes intentional?

**Review by TheRealFalcon — APPROVED** [review]
LGTM, thanks for the tests!

You made some changes to integration_settings.py that I'm assuming weren't intentional, but once those are removed, +1.

File: tests/integration_tests/integration_settings.py (hunk 1 above)
Comment (holmanb) [reply tests/integration_tests/integration_settings.py]: nope, thanks!

**Comment by holmanb** [comment]
> LGTM, thanks for the tests!
> 
> You made some changes to integration_settings.py that I'm assuming weren't intentional, but once those are removed, +1.

Thanks for the review @TheRealFalcon!
